core/src/main/java/com/linecorp/armeria/client/limit/AsyncConcurrencyLimit.java

```diff
@@ -0,0 +1,230 @@
+            // Acquire immediately.
+            acquiredPermitCount.incrementAndGet();
+            promise.setSuccess(semaphorePermit);
+        } else {
+            if (pendingAcquireCount >= maxPendingAcquires) {
+                tooManyOutstanding(promise);
+            } else {
+                final ScheduledFuture<?> timeoutFuture = executor.schedule(timeoutTask,
+                                                                           acquireTimeoutNanos,
+                                                                           TimeUnit.NANOSECONDS);
+                final AcquireTask task = new AcquireTask(executor, promise, timeoutFuture);
+                pendingAcquireQueue.offer(task);
+                ++pendingAcquireCount;
+            }
+
+            assert pendingAcquireCount > 0;
+        }
+    }
+
+    private void tooManyOutstanding(Promise<?> promise) {
+        promise.setFailure(new IllegalStateException("Too many outstanding acquire operations"));
+    }
+
+    private void decrementAndRunTaskQueue() {
+        // We should never have a negative value.
+        final int currentCount = acquiredPermitCount.decrementAndGet();
+        assert currentCount >= 0;
+
+        // Run the pending acquire tasks before notify the original promise so if the user would
+        // try to acquire again from the ChannelFutureListener and the pendingAcquireCount is >=
+        // maxPendingAcquires we may be able to run some pending tasks first and so allow to add
+        // more.
+        runTaskQueue();
+    }
+
+    private void runTaskQueue() {
+        while (acquiredPermitCount.get() < maxParallelism) {
+            final AcquireTask task = pendingAcquireQueue.poll();
+            if (task == null) {
+                break;
+            }
+
+            // Cancel the timeout if one was scheduled
+            final ScheduledFuture<?> timeoutFuture = task.timeoutFuture;
+            timeoutFuture.cancel(false);
+
+            --pendingAcquireCount;
+            task.acquired();
+
+            newPermit(task.promise);
+        }
+
+        // We should never have a negative value.
+        assert pendingAcquireCount >= 0;
+        assert acquiredPermitCount.get() >= 0;
+    }
+
+    private Future<SafeCloseable> newPermit(Promise<SafeCloseable> promise) {
+        promise.setSuccess(semaphorePermit);
+        return promise;
+    }
+
+    private final class AcquireTask implements FutureListener<SafeCloseable> {
+        final Promise<SafeCloseable> promise;
+        final long expireNanoTime = System.nanoTime() + acquireTimeoutNanos;
+        private final EventLoop executor;
+        final ScheduledFuture<?> timeoutFuture;
+        private boolean acquired;
+
+        AcquireTask(EventLoop executor, Promise<SafeCloseable> promise, ScheduledFuture<?> timeoutFuture) {
+            this.executor = executor;
+            this.timeoutFuture = timeoutFuture;
+            this.promise = promise.addListener(this);
```

Comment:
IIUC, we don't really need to add a listener to `promise`. We can just do the necessary action when completing `promise`. For example:

```java
// When acquiring:
acquiredPermitCount.incrementAndGet();
task.promise.trySuccess(semaphorePermit);

// When failed to acquire:
task.promise.tryFailure(exception);
runTaskQueue();
```